docs/operational_guide/bootstrapping.md

```diff
@@ -0,0 +1,106 @@
+
+### Commitlog Bootstrapper
+
+The commitlog bootstrapper's responsibility is to read the commitlog and snapshot (compacted commitlogs) files on disk and recover any data that has not yet been written out as an immutable fileset file. Unlike the filesystem bootstrapper, the commit log bootstrapper cannot simply check which files are on disk in order to determine if it can satisfy a bootstrap request. Instead, the commitlog bootstrapper determines whether it can satisfy a bootstrap request using a simple heuristic.
+
+On a shard-by-shard basis, the commitlog bootstrapper will consult the cluster topology to see if the host it is running on has ever achieved the "Available" status for the specified shard. If so, then the commit log bootstrapper should have all the data since the last fileset file was flushed and will return that it can satisfy any time range for that shard. In other words, the commit log bootstrapper is all-or-nothing for a given shard: it will either return that it can satisfy any time range for a given shard or none at all. In addition, the commitlog bootstrapper *assumes* it is running after the filesystem bootstrapper. M3DB will not allow you to run with a configuration where the filesystem bootstrapper is placed after the commitlog bootstrapper, but it will allow you to run the commitlog bootstrapper without the filesystem bootstrapper which can result in loss of data, depending on the workload.
+
+### Peers Bootstrapper
+
+The peer bootstrapper's responsibility is to stream in data for shard/ranges from other M3DB nodes (peers) in the cluster. This bootstrapper is only useful in M3DB clusters with more than a single node *and* where the replication factor is set to a value larger than 1. The peers bootstrapper will determine whether or not it can satisfy a bootstrap request on a shard-by-shard basis by consulting the cluster topology and determining if there are enough peers to satisfy the bootstrap request. For example, imagine the following M3DB topology where node 1 is trying to perform a peer bootstrap:
+
+| Node | Shard | State |
+|------|-------|-------|
+|  1   |   1   |   I   |
+|  1   |   2   |   I   |
+|  1   |   3   |   I   |
+|  2   |   1   |   I   |
+|  2   |   2   |   I   |
+|  2   |   3   |   I   |
+|  3   |   1   |   A   |
+|  3   |   2   |   A   |
+|  3   |   3   |   A   |
+
+In this case, the peer bootstrapper running on node 1 will not be able to fullfill any requests because node 2 is in the Initializing state for all of its shards and cannot fulfill bootstrap requests. This means that node 1's peer bootstrapper cannot meet its default consistency level of majority for bootstrapping (1 < 2 which is majority with a replication factor of 3). On the other hand, node 1 would be able to peer bootstrap in the following topology because its peers (nodes 2/3) are available for all of their shards:
```

Comment:
Is this case even possible? like 2 replicas in `Initializing` state and 1 in `Available` state?  Also whenever there is a shard in `I` state, there should be a shard in `L` state, can we have that reflected in the examples if possible, I feel it's helpful for users to get a better picture of how the topology looks.

Reply:
yeah its totally possible to have 2 replicas in Initializing and 1 in Available and its not necessarily true that whenever there is a shard in I state there is a corresponding shard in L state (for example, when creating a new cluster)